preview locally using Docker

samzong opened the pull request:
Signed-off-by: samzong.lu <samzong.lu@gmail.com>

**What type of PR is this?**

<!--
Add one of the following kinds:

/kind bug
/kind cleanup
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake

-->

/kind feature

**What this PR does / why we need it**:

a simple way to preview locally using Docker. With just one command, you can easily achieve this.

**Which issue(s) this PR fixes**:
Fixes #181
Fixes #430

**Special notes for your reviewer**:

SAMZONG commented:
/assign @kevin-wangzefeng

RainbowMango commented on .dockerignore:
```diff
@@ -0,0 +1 @@
+**/node_modules
```
Please and a blank line to the end.
Same as below.

RainbowMango reviewed (commented):
Thanks. Please @Poor12 take a look and try it.

SAMZONG commented:
> I have been tried in my local env. Seems there is an err.
> 
> ```
> if ! docker inspect karmada-website:v0.1.0 >/dev/null 2>&1; then \
>         echo "Docker image karmada-website:v0.1.0 does not exist"; \
>         echo "Building Docker image..."; \
>         make image; \
>     fi
> Docker image karmada-website:v0.1.0 does not exist
> Building Docker image...
> make[1]: Entering directory '/root/website'
> docker build -t karmada-website:v0.1.0 .
> Sending build context to Docker daemon  83.71MB
> Step 1/9 : FROM node:16-alpine
> 16-alpine: Pulling from library/node
> 31e352740f53: Pull complete
> fa670e2ca246: Pull complete
> 5307f402fec1: Pull complete
> e3da03619ad7: Pull complete
> Digest: sha256:296f32946d2fb5227e43fa51e608bbb11a208d648a833e9b264934fb0c7f404b
> Status: Downloaded newer image for node:16-alpine
>  ---> 4874253f493b
> Step 2/9 : WORKDIR /app/docs
>  ---> Running in 0bf40b1ca71d
> Removing intermediate container 0bf40b1ca71d
>  ---> b3bf4b3636b7
> Step 3/9 : COPY package.json package-lock.json ./
>  ---> 29b2fcdc8061
> Step 4/9 : RUN npm install
>  ---> Running in 83ecd45b4c3c
> npm WARN deprecated trim@0.0.1: Use String.prototype.trim() instead
> npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
> npm notice
> npm notice New major version of npm available! 8.19.4 -> 9.7.1
> npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.7.1>
> npm notice Run `npm install -g npm@9.7.1` to update!
> npm notice
> npm ERR! code ERR_SOCKET_TIMEOUT
> npm ERR! network Socket timeout
> npm ERR! network This is a problem related to network connectivity.
> npm ERR! network In most cases you are behind a proxy or have bad network settings.
> npm ERR! network
> npm ERR! network If you are behind a proxy, please make sure that the
> npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /root/.npm/_logs/2023-06-19T06_48_47_373Z-debug-0.log
> The command '/bin/sh -c npm install' returned a non-zero code: 1
> make[1]: *** [Makefile:20: image] Error 1
> make[1]: Leaving directory '/root/website'
> make: *** [Makefile:6: serve] Error 2
> ```

@Poor12

It's seems to be an issue with the default npm registry timeout on your computer,

```bash
npm ERR! code ERR_SOCKET_TIMEOUT
npm ERR! network Socket timeout
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
```

Poor12 commented:
Sorry, I cannot fix it. Could you show your test results?

SAMZONG commented:
@Poor12 

this is my screen.

[![asciicast](https://asciinema.org/a/83T4IaGm1pHFVLLBdKclLQp7k.svg)](https://asciinema.org/a/83T4IaGm1pHFVLLBdKclLQp7k)

Poor12 commented:
I first run `make image` and then I run `make serve`. It works well. Must I run `make image` before `make serve`? If so, I suggest adding it to the doc.

SAMZONG commented:
> I first run `make image` and then I run `make serve`. It works well. Must I run `make image` before `make serve`? If so, I suggest adding it to the doc.


this not necessary, when running `make serve`, it will check if image 
doesn't, it's will automatically build it.

You can confirm this from the video in above comment.

Poor12 commented on bs-config.js:
```diff
@@ -0,0 +1,13 @@
+    // Close external access.
+    open: false,
+  };
+  
```
Add a blank line.

Poor12 commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+		echo "Docker image $(IMAGE_NAME):$(IMAGE_VERSION) does not exist"; \
+		echo "Building Docker image..."; \
+		make image; \
+    fi
```
Misaligned.

Poor12 reviewed (commented):
LGTM
Just so typos.

SAMZONG commented:
next actions:

1. Image optimization
2. Pushing the image to a public repository, so that users do not need to build it themselves
3. Adding more functionality, such as automatic markdownlint
4. ...

SAMZONG commented:
@RainbowMango @Poor12 Is there any issue with this PR? shall we merge it?

SAMZONG commented:
If having some question, we can about next meeting.

samzong commented:
What did this PR update?

- Updated the base image to nodejs v18.
- rebase to lastest branch
- Refactored README.md

Because we are still using `Yarn` for production builds, it is recommended that we indicate this in the README and provide corresponding operation instructions.

@windsonsea can u help me review README.md ?

windsonsea reviewed (commented):
Thanks. Here are some text suggestions in README for fluency reference.

windsonsea commented:
Those text in README looks good to me.
/lgtm

samzong commented:
> What did this PR update?
> 
> * Updated the base image to node.js v18.
> * rebase to lastest branch
> * Refactored README.md
> 

Because we are still using `Yarn` for production builds, it is recommended that we indicate this in the README and provide corresponding operation instructions.

---

An outdated PR, let's continue to move forward.

@RainbowMango @Arhell I have updated the PR, and pls looking forward to review.

thanks.

Arhell reviewed (commented):
lgtm

RainbowMango reviewed (approved):
/lgtm
/approve

Thanks.
